modules/cudaarithm/src/cuda/threshold.cu

```diff
@@ -95,12 +95,232 @@ namespace
+    {
+        variance[threshold] = make_longlong2(threshold_variance_above_i64, threshold_variance_below_i64);
+    }
+}
+
+template <uint n_thresholds>
+__global__ void
+otsu_score(uint *otsu_threshold, uint *threshold_sums, longlong2 *variance)
+{
+    extern __shared__ float shared_memory_f32[];
+
+    uint threshold = blockIdx.x * blockDim.x + threadIdx.x;
+
+    uint n_samples = threshold_sums[0];
+    uint n_samples_above = threshold_sums[threshold];
+    uint n_samples_below = n_samples - n_samples_above;
+
+    float threshold_mean_above = (float)n_samples_above / n_samples;
+    float threshold_mean_below = (float)n_samples_below / n_samples;
+
+    longlong2 variances = variance[threshold];
+    float variance_above = (float)variances.x / n_samples_above;
+    float variance_below = (float)variances.y / n_samples_below;
+
+    float above = threshold_mean_above * variance_above;
+    float below = threshold_mean_below * variance_below;
+    float score = above + below;
+
+    float original_score = score;
+
+    blockReduce<n_thresholds>((float *)shared_memory_f32, score, threshold, minimum<float>());
+
+    if (threshold == 0)
+    {
+        shared_memory_f32[0] = score;
+    }
+    __syncthreads();
+
+    score = shared_memory_f32[0];
+
+    // We found the minimum score, but we need to find the threshold. If we find the thread with the minimum score, we
+    // know which threshold it is
+    if (original_score == score)
+    {
+        *otsu_threshold = threshold - 1;
+    }
+}
+
+template <uint n_bins, uint n_thresholds>
+void compute_otsu_async(uint *histogram, uint *otsu_threshold, Stream &stream)
+{
+    CV_Assert(n_bins == 256);
+    CV_Assert(n_thresholds == 256);
+
+    cudaStream_t cuda_stream = StreamAccessor::getStream(stream);
+
+    dim3 block_all = dim3(n_bins, 1, 1);
+    dim3 grid_all = dim3(divUp((int)n_bins, block_all.x), divUp((int)n_thresholds, block_all.y), 1);
+    dim3 block_score = dim3(n_thresholds, 1, 1);
+    dim3 grid_score = dim3(divUp((int)n_thresholds, block_score.x), 1, 1);
+
+    GpuMat gpu_threshold_sums = GpuMat(1, n_bins, CV_32SC1);
+    GpuMat gpu_sums = GpuMat(1, n_bins, CV_64FC1);
+    GpuMat gpu_variances = GpuMat(1, n_bins, CV_32SC4);
+
+    uint shared_memory;
+    shared_memory = n_bins * sizeof(unsigned long long);
+    otsu_mean<256><<<grid_all, block_all, shared_memory, cuda_stream>>>(histogram, gpu_threshold_sums.ptr<uint>(), gpu_sums.ptr<unsigned long long>());
+    otsu_variance<256><<<grid_all, block_all, shared_memory, cuda_stream>>>(gpu_variances.ptr<longlong2>(), histogram, gpu_threshold_sums.ptr<uint>(), gpu_sums.ptr<unsigned long long>());
+
+    shared_memory = n_bins * sizeof(float);
+    otsu_score<256><<<grid_score, block_score, shared_memory, cuda_stream>>>(
+        otsu_threshold, gpu_threshold_sums.ptr<uint>(), gpu_variances.ptr<longlong2>());
+}
+
+// TODO: Replace this is cv::cuda::calcHist
+template <uint n_bins>
+__global__ void histogram_kernel(
```

Comment:
@asmorkalov Due to a circular dependency when adding the **opencv_cudaimgproc** to **opencv_cudaarithm** in _CMakeLists.txt_ `cv::cuda::calcHist` cannot be called from here.  To avoid refactoring would it make sence at a minimum to only have a single implementation of `calcHist` even if it is duplicated or do you have a better suggestion? i.e. copy the implemntation from  `cv::cuda::calcHist` here.

Reply:
@asmorkalov any thoughts on this?